Add initial attestation section. (#36)
riscv-non-isa/riscv-ap-tee-io: specification/08-attestation.adoc

[[attestation]]
== Device Attestation

There are two attestation stages involved in the CoVE-IO architecture.

> - TVM attests the device

Reviewer: Nit: You need a breaking line here or else it won't be nicely formatted into 2 separare points.

Author: yes. Done

architecture, because it adds complexity to the TSM as the TSM would need to
perform the verification and pass the attestation result to TVM.
Since the TSM is in the TVM's TCB, its attack surface should be kept minimal.

In step 3, the verifier needs to perform the verification for the device. There
are two possible use cases:

==== Local Verification
With local verification, the endorsement and reference values are provisioned
into the TVM. 
  
Static provisioning means the data is built into the TVM image at build time,
while dynamic provisioning means the data is provisioned into the TVM at runtime.

The endorsement may include the device root CA certificate and its associated
certificate revocation list (CRL). The reference value includes the reference
integrity manifest (RIM), also known as golden measurement.

When running locally, the verifier in the TVM validates the device certificate
according to the device root CA certificate and CRL, and then compares the
device measurement with the measurement from its locally provisioned RIM.

==== Remote Verification
When the device is verified remotely, the endorsement and reference values

> are not provisioned into the TVM, but are accessible from on an externally

Reviewer: Typo:

```suggestion
are not provisioned into the TVM, but are accessible from an externally
```

Author: done

available service, also known as a relying party for the device.

The verifier itself can either run locally or be hosted on an online service
as well, typically available through an HTTPS-based protocol.

> It always collect the up to date endorsement and RIM from the device relying party.

Reviewer: Typo:

```suggestion
It always collects the up-to-date endorsement and RIM from the device relying party.
```

Author: done

[[attestation]]
== Device Attestation

> There are two attestation involved in the CoVE-IO architecture.

Author: How about we change to `two attestation stages`?

I have concern on using `model`, because `two models` may make people there are 2 possibilities and they can choose 1 of them.

> There are two attestation involved in the CoVE-IO architecture.

Reviewer: Makes sense to me, yes.

> There are two attestation involved in the CoVE-IO architecture.

Author: done

> - TVM attests the Device

Author: done

> - 3rd party attests the TVM with the device

Reviewer: ```suggestion
- A third party attests the TVM with the device
```

Author: done

The IETF RATS Remote Attestation Architecture <<RATS>> can be applied to both.

.Remote Attestation Framework (IETF RATS)
image::images/rats-arch.svg[align="center"]

=== TVM attesting the Device

> The TVM needs to verify and trust the device before offload the workload to the

Reviewer: ```suggestion
The TVM must verify and trust the device before offloading the workload to the
```

Author: done

> TDI. According to <<TDISP>> architecture, the high level steps are as follows:

Reviewer: ```suggestion
TDI. According to the <<TDISP>> architecture, the high level steps for that model are as follows:
```

Author: done

1. The TSM collects the device certificate and measurement from the DSM as the
   evidence, via SPDM protocol.
2. The TSM presents the evidence to the TVM.
3. A TVM verifier appraises the evidence and produce the attestation result to

>    the relying party in the TVM, such as device consumer.

Reviewer: What do you mean by "such as device consumer" ?

Author: I mean the code that uses the device, such as vendor's device driver, or generic PCI driver.

> 4. The relying part in the TVM accepts the device into its TCB.

Reviewer: ```suggestion
4. The local relying part running in the TVM accepts (or rejects) the device into (from) its TCB.
```

Author: done

> Above process follows the background-check model according to <<RATS>>.

Reviewer: ```suggestion
The above process follows the background-check model, as defined by <<RATS>>.
```

Author: done

> We do not consider the passport model, because that adds complexity to the TSM

Reviewer: ```suggestion
The <<RATS>> passport model is not supported by the CoVE-IO attestation architecture, because it adds complexity to the TSM
```

Author: done

> if the TSM needs to perform the verification and pass the attestation result to

Reviewer: ```suggestion
as the TSM would need to perform the verification and pass the attestation result to
```

Author: done

> TVM. Since TSM is TCB of TVM, the TSM should be kept as minimal.

Reviewer: ```suggestion
TVM. Since the TSM is in the TVM's TCB, its attack surface should be kept minimal.
```

In step 3, the verifier needs to perform the verification for the device. There
are two possible use cases:

> - Local Verification. The endorsement and reference value are provisioned inside
>   of TVM. Static provisioning means the data is built into TVM image, while
>   dynamic provisioning means the data is input to TVM at runtime.
>   The endorsement may include device root CA certificate and device certificate
>   revocation list (CRL). The reference value includes the reference integrity
>   manifest (RIM), also known as golden measurement.
>   The verifier in the TVM may just validate the device certificate according to
>   the device root CA certificate and CRL, then compare the device measurement
>   with the measurement in the RIM locally.

Reviewer: ```suggestion
==== Local Verification
With local verification, the endorsement and reference values are provisioned into the TVM. 
  
Static provisioning means the data is built into the TVM image at build time, while dynamic provisioning
means the data is provisioned into the TVM at runtime.

The endorsement may include the device root CA certificate and its associated certificate
revocation list (CRL). The reference value includes the reference integrity manifest (RIM),
also known as golden measurement.

When running locally, the verifier in the TVM validates the device certificate
according to the device root CA certificate and CRL, and then compares the
device measurement with the measurement from its locally provisioned RIM.
```

Author: done

> - Remote Verification. The endorsement and reference value are not provisioned
>   in the TVM, but only available in a device service on the web.
>   The verifier could be inside of the TVM, or a device verification service on
>   the web. The verifer always collect the up to date endorsement and RIM from
>   the device service.
> 

Reviewer: ```suggestion
==== Remote Verification
When the device is verified remotely, the endorsement and reference values
are not provisioned into the TVM, but are accessible from on an externally
available service, also known as a relying party for the device.

The verifier itself can either run locally or be hosted on an online service
as well, typically available through an HTTPS-based protocol.
It always collect the up to date endorsement and RIM from the device relying party.
```

Author: done

.Device Attestation Comparison
[width=90%, align="center", options="header"]
|===
| Use Case                        | Local Verification                     | Remote Verification
| Endorsement (Root CA Cert, CRL) | Provisioned in TVM (Static or Dynamic) | Online Service
| Reference Value (RIM)           | Provisioned in TVM (Static or Dynamic) | Online Service
| Verifier Location               | Inside of TVM         | Inside of TVM or Online Service
|===

> === 3rd party attesting the TVM with the device

Reviewer: ```suggestion
=== Third party attesting the TVM with the device
```

> The 3rd party need follow the CoVE defined attestation mechanism to perform

Reviewer: ```suggestion
The 3rd party must follow the CoVE defined attestation mechanism to perform
```

Author: done

> local or remote attestation and verify TVM.

Reviewer: ```suggestion
local or remote attestation and verify the TVM.
```

Author: done

The CoVE-IO architecture adds the additional device related information to the

> TVM evidence, to allow the 3rd party to verify the attached device TDI.

Reviewer: ```suggestion
TVM attestation evidence, in order to allow the 3rd party to verify the attached device TDI.
```

Author: done

> The TVM needs to add different device information to the TVM report based on

Reviewer: ```suggestion
Depending on the verification process being local or remote, the TVM needs to add different pieces of device-related information to the TVM report:
```

Author: done

different use cases:

> - Local Verification. The TVM report should include the device verifier code,
>   the device policy, including the provisioned device root CA cert, CRL and RIM.
>   If the TVM supports device policy update, then the secure signed update
>   mechanism should be implemented. In such case, the signer of the update data
>   shall be included in the TVM report as the trust anchor.
>   The device measurement and certificate are not required in the TVM report,
>   but the TVM should provide a mechanism to return the device measurement and
>   certificate to the verifier for furthur verification.

Reviewer: ```suggestion
==== Local Verification. 

The TVM report should include the device verifier code, the device policy, including the provisioned device root CA certificate, CRL and RIM.

If the TVM supports device policy update, then the secure signed update mechanism should be implemented. In that case, the signer of the update data shall be included in the TVM report as the trust anchor.

The device measurement and certificate are not required in the TVM report, however the TVM should provide a mechanism to return the device measurement and certificate to the verifier for further verification.
```

Author: done

> - Remote Verification. The TVM report should include the device verifier code
>   or the stub code connecting the verifier service. The TVM report should
>   include the online service URL and the public certificate as the trust anchor.
>   The detailed up to date endorsement or reference value is NOT required be in
>   the TVM report. 
>   The device measurement and certificate are not required in the TVM report,
>   but the TVM should provide a mechanism to return the device measurement and
>   certificate to the verifier for furthur verification.

Reviewer: ```suggestion
==== Remote Verification

The TVM report should include measurements of either the device verifier or the proxy connecting to the verifier service. It should also include the online service URL and the public certificate as the trust anchor.

The detailed, up-to-date endorsement or reference value is *not* required to be part of the TVM report. 

Although the device measurement and certificate are not required to be included in the TVM report, the TVM should provide a mechanism to return the device measurement and certificate for the verifier to perform further verification.
```

Author: done
